ui-admin/src/study/surveys/SurveyView.tsx

```diff
@@ -53,30 +54,62 @@ function RawSurveyView({ studyEnvContext, survey, readOnly = false }:
       readOnly={readOnly}
-      onCancel={() => navigate(currentEnvPath)}
+      onCancel={() => navigate(studyEnvFormsPath(portal.shortcode, study.shortcode, currentEnv.environmentName))}
       onSave={createNewVersion}
     />
   )
 }
 
-/** routable component for survey editing */
-function SurveyView({ studyEnvContext }: {studyEnvContext: StudyEnvContextT}) {
-  const params = useParams<SurveyParamsT>()
-  const surveyStableId: string | undefined = params.surveyStableId
+/** loads a survey and associated data (e.g. answer mappings) */
+export const useLoadedSurvey = (portalShortcode: string, stableId: string, version: number) => {
+  const [survey, setSurvey] = useState<Survey | undefined>()
+  const [isLoading, setIsLoading] = useState(true)
+
+  /** load the survey from the server to get answer mappings and ensure we've got the latest content */
+  useEffect(() => {
+    setIsLoading(true)
+    Api.getSurvey(portalShortcode, stableId, version).then(result => {
+      setSurvey(result)
+      setIsLoading(false)
+    }).catch(e => {
+      Store.addNotification(failureNotification(`could not load survey ${  e.message}`))
+    })
+  }, [portalShortcode, stableId, version])
 
-  const { currentEnv } = studyEnvContext
+  return { isLoading, survey }
+}
+
+/** read survey-related url params */
+export const useSurveyParams = () => {
+  const params = useParams<SurveyParamsT>()
   const [searchParams] = useSearchParams()
   const isReadOnly = searchParams.get('readOnly') === 'true'
+  const version = params.version ? parseInt(params.version) : undefined
 
-  if (!surveyStableId) {
-    return <span>you need to specify the stableId of the survey</span>
-  }
+  return { isReadOnly, version, stableId: params.surveyStableId }
+}
 
-  const survey = currentEnv.configuredSurveys.find(s => s.survey.stableId === surveyStableId)?.survey
-  if (!survey) {
-    return <span>The survey {surveyStableId} does not exist in this environment</span>
+/** routable component for survey editing */
+function SurveyView({ studyEnvContext }: {studyEnvContext: StudyEnvContextT}) {
+  const { isReadOnly, version, stableId } = useSurveyParams()
+  const { currentEnv, portal } = studyEnvContext
+
+  const envSurvey = currentEnv.configuredSurveys
+    .find(s => s.survey.stableId === stableId)?.survey
+  const appliedVersion = version || envSurvey?.version
+  if (!stableId) {
+    return <span>You must specify a stableId for the survey to edit</span>
+  }
+  if (!appliedVersion) {
+    return <span>The survey {stableId} is not already configured for this environment
+      -- you must specify a version</span>
   }
-  return <RawSurveyView studyEnvContext={studyEnvContext} survey={survey} readOnly={isReadOnly}/>
+  const { isLoading, survey } = useLoadedSurvey(portal.shortcode, stableId, appliedVersion)
+
+  return <>
+    { isLoading && <LoadingSpinner/> }
+    { !isLoading && <RawSurveyView studyEnvContext={studyEnvContext} survey={survey!} readOnly={isReadOnly}/> }
```

Comment:
I'm mainly asking this for my own TypeScript learning: the non-null assertion on `survey` looks safe because the `useLoadedSurvey` implementation seems to guarantee that `isLoading` will never be `false` if `survey` is `undefined`, but is there a reason to not do the below suggestion instead? Is this a case where we view TypeScript more as a tool and don't want it to dictate/get in the way too much if we know better than it does? I'm still trying to dial-in on when circumventing TypeScript checks are appropriate.

```suggestion
    { !isLoading && survey && <RawSurveyView studyEnvContext={studyEnvContext} survey={survey} readOnly={isReadOnly}/> }
```

Reply:
good question, I could see that going either way in this case